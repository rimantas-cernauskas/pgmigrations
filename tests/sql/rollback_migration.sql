-- Test function pgmigrations.rollback_migration
\i tests/sql/setup/init.sql
BEGIN;
SELECT plan(9);
SET ROLE __pgmigrations_testuser;

-- Setup migration scripts
SET ROLE __pgmigrations_superuser;
SELECT is(
    pgmigrations.add_migration(
         '00'
        ,'CREATE TABLE table0( id SERIAL );'
        ,'DROP TABLE table0;'
    )
    ,TRUE
    ,'should add migration "00"'
);

-- Test rollback_migrations() error
SET ROLE __pgmigrations_testuser;
SELECT is(
     pgmigrations.rollback_migration('00')
    ,FALSE
    ,'rollback_migration() should return FALSE on unsuccessful run'
);

PREPARE my_query AS
    SELECT
         migration
        ,down_run
        ,down_success
        ,down_error
    FROM pgmigrations.migrations_info
    ORDER BY migration ASC;

PREPARE my_expect AS
    SELECT * FROM (
        VALUES (
             '00'
            ,NOW()
            ,FALSE
            ,'table "table0" does not exist'
        )
    )
    t(
         migration
        ,down_run
        ,dwon_success
        ,down_error
    )
;

SELECT results_eq(
     'my_query'
    ,'my_expect'
    ,'should get expected migrations info'
);
DEALLOCATE my_query;
DEALLOCATE my_expect;


-- apply migration, so it could be rolled back
SELECT is(
     pgmigrations.apply_migrations()
    ,TRUE
    ,'apply_migrations() should return TRUE'
);

-- check that migrations were actually applied
-- migration 00
SELECT has_table('table0');
SELECT has_column('table0', 'id');

-- rollback migration
SELECT is(
    pgmigrations.rollback_migration(
        '00'
    )
    ,TRUE
    ,'rollback_migration() should return TRUE on successful rollback'
);

-- check log
PREPARE my_query AS
    SELECT
         migration
        ,down_run
        ,down_success
        ,down_error
    FROM pgmigrations.migrations_info
    ORDER BY migration ASC;

PREPARE my_expect AS
    SELECT * FROM (
        VALUES (
             '00'
            ,NOW()
            ,TRUE
            ,NULL
        )
    )
    t(
         migration
        ,down_run
        ,dwon_success
        ,down_error
    )
;

SELECT results_eq(
     'my_query'
    ,'my_expect'
    ,'should get expected migrations info'
);
DEALLOCATE my_query;
DEALLOCATE my_expect;

-- check that table has been deleted
SELECT hasnt_table('table0');

SELECT * FROM finish();
ROLLBACK;