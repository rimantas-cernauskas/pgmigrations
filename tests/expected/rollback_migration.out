-- Test function pgmigrations.rollback_migration
\i tests/sql/setup/init.sql
\set ECHO errors
1..9
ok 1 - should add migration "00"
ok 2 - rollback_migration() should return FALSE on unsuccessful run
ok 3 - should get expected migrations info
ok 4 - apply_migrations() should return TRUE
ok 5 - Table table0 should exist
ok 6 - Column table0.id should exist
ok 7 - rollback_migration() should return TRUE on successful rollback
ok 8 - should get expected migrations info
ok 9 - Table table0 should not exist
